### PR TITLE
API refactorings, `transactions` module overhaul

### LINKED ISSUE #71: Impl Display and FromStr for RevaultTransaction
As for serde, just use as_psbt_string and from_psbt_str

### LINKED ISSUE #81: fuzz: fix run.sh script and some warnings in fuzz_targets
It's not clear to know where to use run.sh, `./fuzz/run.sh` or `cd fuzz & ./run.sh` ?
```bash
cd corpus && git clone https://github.com/revault/revault_tx_corpus

cargo install --force cargo-fuzz
for target in $(ls fuzz/fuzz_targets);do
    cargo +nightly fuzz run "${target%.*}" -- -runs=0
done
```
corpus is in `fuzz` directory and `for target in $(ls fuzz/fuzz_targets)` means we are in `revault_tx` root directory
 
---

When running some warnings are often:
```
warning: unused `Result` that must be used
  --> fuzz_targets/parse_unvault.rs:45:9
   |
45 |         tx.verify_input(0);
   |         ^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_must_use)]` on by default
   = note: this `Result` may be an `Err` variant, which should be handled

warning: unused `Result` that must be used
  --> fuzz_targets/parse_unvault.rs:51:9
   |
51 |         tx.finalize(&secp);
   |         ^^^^^^^^^^^^^^^^^^^
   |
   = note: this `Result` may be an `Err` variant, which should be handled

```

### LINKED ISSUE #85: Commit to a MSRV
1.43 looks reasonable (1.48 in Debian bullseye), need to dig more into `rustc`'s release notes.

### PR DESCRIPTION
**Pull request by darosior**
This is a set of refactorings and API improvements. In a nutshell this:
- Introduces a type separation for our 2 differents kind of outputs
- Fixes some long-standing issues with the API (`signature_hash`, `verify_inputs`, ..)
- Adds some more wrappers as we appear to need them downstream
- Some parsing checks additions and fixes

I unfortunately couldn't find a way to get around https://github.com/revault/revault_tx/issues/57 .

I also didn't address https://github.com/revault/revault_tx/issues/79 as i think a `RevocationTransaction` trait is unnecessary. However i'm now leaning toward introducing finalized transactions into the type system, we'll see.

Fixes #71
Fixes #81
Fixes #85

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,6 +6,10 @@ Bitcoin vault architecture for multi-party situations.
 This library defines the Miniscript descriptors and handles PSBT management of
 Revault-specific transactions.
 
+# Minimum Supported Rust Version
+
+This library should always compile with any combination of features on **Rust 1.43**.
```
Comment (danielabrozzoni) [inline README.md:11]: 🎉

File: src/transactions/utils.rs [hunk 2]
```diff
@@ -170,6 +166,26 @@ pub fn psbt_common_sanity_checks(psbt: Psbt) -> Result<Psbt, PsbtValidationError
...
+        if spk.is_v0_p2wsh() {
+            // It's blanked when finalized
+            if is_final == Some(true) {
+                continue;
+            }
+
+            let ws = match &input.witness_script {
+                Some(ws) => ws,
+                None => return Err(PsbtValidationError::MissingInWitnessScript(input.clone())),
+            };
```
Comment (danielabrozzoni) [inline src/transactions/utils.rs:182]:
```
            let ws = input
                .witness_script
                .as_ref()
                .ok_or_else(|| PsbtValidationError::MissingInWitnessScript(input.clone()))?;
```

File: src/transactions/mod.rs [hunk 3]
```diff
@@ -73,15 +73,15 @@ pub const MAX_STANDARD_TX_WEIGHT: u32 = 400_000;
 /// - Finalizer
 /// - Extractor and serializer
 pub trait RevaultTransaction: fmt::Debug + Clone + PartialEq {
-    /// Get the inner transaction
-    fn inner_tx(&self) -> &Psbt;
+    /// Get the inner PSBT
+    fn psbt(&self) -> &Psbt;
```
Comment (danielabrozzoni) [inline src/transactions/mod.rs:78]: +1

**Review by danielabrozzoni — COMMENTED** [review]
I reviewed briefly, the API looks good to me

**Comment by darosior** [comment]
Found a nice overlook when integrating it in `revaultd`:
```diff
diff --git a/src/transactions/cancel.rs b/src/transactions/cancel.rs
index 0ff4a33..3a53b17 100644
--- a/src/transactions/cancel.rs
+++ b/src/transactions/cancel.rs
@@ -141,9 +141,6 @@ impl CancelTransaction {
 
         // Deposit txo is P2WSH
         let output = &psbt.outputs[0];
-        if output.witness_script.is_none() {
-            return Err(PsbtValidationError::MissingOutWitnessScript(output.clone()).into());
-        }
         if output.redeem_script.is_some() {
             return Err(PsbtValidationError::InvalidOutputField(output.clone()).into());
         }
diff --git a/src/transactions/mod.rs b/src/transactions/mod.rs
index 076d1a1..a7c0711 100644
--- a/src/transactions/mod.rs
+++ b/src/transactions/mod.rs
@@ -1252,6 +1252,7 @@ mod tests {
                 };
             }
 
+            roundtrip!(cancel_tx);
             roundtrip!(emergency_tx);
             roundtrip!(unvault_tx);
             roundtrip!(unemergency_tx);
```

**Comment by darosior** [comment]
Pushed a commit based on feedback from using this API for the watchtower.

**Comment by darosior** [comment]
ACK 7cc82a0

**Pull request merged**